Added sample models

shehadak opened the pull request:
This PR adds a sample of neural network models including DenseNet, EfficientNet, GoogLeNet, Inception v3, ResNet, and Vision Transformer using the Brain-Score 2.0 format. This PR aims to get users up to speed for testing their benchmarks on a series of models and provides example models for 

The full list of models is being added in #498.

mschrimpf commented on brainscore_vision/models/alexnet/model.py:
```diff
@@ -1,28 +1,41 @@
 
 def get_model():
-    model = torchvision.models.alexnet(pretrained=True)
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
I think this should rather be in the wrapper/activations helper (which has to also move stimuli to gpu)

shehadak replied:
You're correct. It already exists in the Pytorch wrapper, so no need for it here.

mschrimpf commented on brainscore_vision/models/alexnet/test.py:
```diff
@@ -2,12 +2,14 @@
 
 from brainscore_vision import score
+from brainscore_vision.utils import seed_everything
+seed_everything(42)
```
move inside methods where needed. `pytest` does not guarantee sequential test execution for instance.

But I also don't see what exactly we have to seed here. Everything should be deterministic:
* the model is uniquely specified
* the benchmarks have no randomness to them

So why do we need to seed?

shehadak replied:
Seeding is necessary for models that generally use randomness during inference (e.g. untrained models with random initialization, VAEs, MC dropout), resulting in different scores even for the same benchmark run. I also think, generally, it's a good practice to seed test running for consistency. None of the models in this PR use it, so I removed seeding for this PR.

mschrimpf commented on brainscore_vision/models/alexnet/test.py:
```diff
@@ -2,12 +2,14 @@
+seed_everything(42)
 
 
 @pytest.mark.private_access
```
```suggestion
```

mschrimpf commented on brainscore_vision/models/alexnet/test.py:
```diff
@@ -2,12 +2,14 @@
 @pytest.mark.memory_intense
 @pytest.mark.parametrize("model_identifier, benchmark_identifier, expected_score", [
-    ("alexnet", "MajajHong2015.IT-pls", approx(0.508, abs=0.0005)),
+    ("alexnet", "MajajHong2015public.IT-pls", approx(0.5048, abs=0.0005)),
```
```suggestion
    pytest.param("alexnet", "MajajHong2015.IT-pls", approx(0.508, abs=0.0005), marks=[pytest.mark.private_access]),
    pytest.param("alexnet", "MajajHong2015public.IT-pls", approx(0.5048, abs=0.0005)),
```

mschrimpf replied:
let's not throw away tests

mschrimpf commented on brainscore_vision/models/crossvit_pretrained/model.py:
```diff
@@ -0,0 +1,62 @@
+    "blocks.1.blocks.1.0.mlp.act",
+    "blocks.2.revert_projs.1.2",
+]
+# Description of Layers:
```
where do these come from?

shehadak replied:
The original model submission (`cv_18_dagger_408_pretrained`) from braintree.

mschrimpf commented on brainscore_vision/models/crossvit_pretrained/model.py:
```diff
@@ -0,0 +1,62 @@
+
+def get_model():
+    # Generate Model
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
as for alexnet, move to wrapper

mschrimpf commented on brainscore_vision/models/crossvit_pretrained/model.py:
```diff
@@ -0,0 +1,62 @@
+    preprocessing = functools.partial(
+        load_preprocess_custom_model, image_size=224)
+    # get an activations model from the Pytorch Wrapper
+    activations_model = PytorchWrapper(
+        identifier="cv_18_dagger_408_pretrained", model=model, preprocessing=preprocessing
+    )
+
+    wrapper = activations_model
```
```suggestion
    wrapper = PytorchWrapper(
        identifier="cv_18_dagger_408_pretrained", model=model, preprocessing=preprocessing
    )
```

mschrimpf replied:
bump

mschrimpf commented on brainscore_vision/models/crossvit_pretrained/test.py:
```diff
@@ -0,0 +1,18 @@
+seed_everything(42)
+
+
+@pytest.mark.private_access
```
```suggestion
```

mschrimpf commented on brainscore_vision/models/crossvit_pretrained/test.py:
```diff
@@ -0,0 +1,18 @@
+
+from brainscore_vision import score
+from brainscore_vision.utils import seed_everything
+seed_everything(42)
```
as above, move into respective functions

mschrimpf commented on brainscore_vision/models/densenet/model.py:
```diff
@@ -0,0 +1,59 @@
+def get_model(net):
+    assert net in net_constructors, f"Could not find DenseNet network: {net}"
+
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
move to wrapper

mschrimpf commented on brainscore_vision/models/densenet/test.py:
```diff
@@ -0,0 +1,31 @@
+from brainscore_vision import score
+from brainscore_vision.utils import seed_everything
+
+seed_everything(42)
```
move into function

mschrimpf commented on brainscore_vision/models/densenet/test.py:
```diff
@@ -0,0 +1,31 @@
+logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
+
+
+@pytest.mark.private_access
```
```suggestion

```

mschrimpf commented on brainscore_vision/models/crossvit_pretrained/test.py:
```diff
@@ -0,0 +1,18 @@
+])
+def test_score(model_identifier, benchmark_identifier, expected_score):
+    actual_score = score(model_identifier=model_identifier, benchmark_identifier=benchmark_identifier,
+                         conda_active=True)
```
I can see having to set this flag every time being confusing for users. @kvfairchild can we simplify this?

mschrimpf commented on brainscore_vision/models/efficientnet/model.py:
```diff
@@ -0,0 +1,53 @@
+def get_model(net):
+    assert net in net_constructors, f"Could not find EfficientNet network: {net}"
+
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
move into wrapper

mschrimpf commented on brainscore_vision/models/efficientnet/test.py:
```diff
@@ -0,0 +1,35 @@
+from brainscore_vision.utils import seed_everything
+
+logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
+seed_everything(42)
```
move into function

mschrimpf commented on brainscore_vision/models/efficientnet/test.py:
```diff
@@ -0,0 +1,35 @@
+seed_everything(42)
+
+
+@pytest.mark.private_access
```
```suggestion
```

mschrimpf commented on brainscore_vision/models/googlenet/model.py:
```diff
@@ -0,0 +1,43 @@
+
+
+def get_model():
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
move into wrapper

mschrimpf commented on brainscore_vision/models/googlenet/test.py:
```diff
@@ -0,0 +1,28 @@
+from brainscore_vision.utils import seed_everything
+
+logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
+seed_everything(42)
```
move into function/delete

mschrimpf commented on brainscore_vision/models/inception_v3/model.py:
```diff
@@ -0,0 +1,48 @@
+
+
+def get_model():
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
move into wrapper

mschrimpf commented on brainscore_vision/models/inception_v3/test.py:
```diff
@@ -0,0 +1,28 @@
+from brainscore_vision.utils import seed_everything
+
+logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
+seed_everything(42)
```
move into function

mschrimpf commented on brainscore_vision/models/inception_v3/test.py:
```diff
@@ -0,0 +1,28 @@
+seed_everything(42)
+
+
+@pytest.mark.private_access
```
```suggestion
```

mschrimpf commented on brainscore_vision/models/pixels/model.py:
```diff
@@ -20,7 +21,8 @@ def identifier(self):
         self._extractor.identifier = value
 
-    def __call__(self, *args, **kwargs):  # cannot assign __call__ as attribute due to Python convention
+    # cannot assign __call__ as attribute due to Python convention
```
```suggestion
    # Note that we cannot assign __call__ as attribute due to Python convention, 
    # so we cannot set `self.__call__ = self._extractor` in the `__init__` and 
    # instead have to explicitly implement the function here.
```

mschrimpf replied:
bump

mschrimpf commented on brainscore_vision/models/pixels/test.py:
```diff
@@ -2,12 +2,14 @@
+seed_everything(42)
 
 
 @pytest.mark.private_access
```
```suggestion
```

mschrimpf commented on brainscore_vision/models/pixels/test.py:
```diff
@@ -2,12 +2,14 @@
 @pytest.mark.memory_intense
 @pytest.mark.parametrize("model_identifier, benchmark_identifier, expected_score", [
-    ("pixels", "MajajHong2015.IT-pls", approx(0.0153, abs=0.0005)),
+    ("pixels", "MajajHong2015public.IT-pls", approx(0.0764, abs=0.0005)),
```
```suggestion
    pytest.param("pixels", "MajajHong2015.IT-pls", approx(0.0153, abs=0.0005), marks=[pytest.mark.private_access]),
    pytest.param("pixels", "MajajHong2015public.IT-pls", approx(0.0764, abs=0.0005)),
```

mschrimpf commented on brainscore_vision/models/pixels/requirements.txt:
```diff
@@ -0,0 +1,2 @@
+torch
+torchvision
```
I don't think we need those here

mschrimpf replied:
bump

mschrimpf commented on brainscore_vision/models/resnet/model.py:
```diff
@@ -0,0 +1,67 @@
+def get_model(net):
+    assert net in net_constructors, f"Could not find ResNet network: {net}"
+
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
move to wrapper

mschrimpf commented on brainscore_vision/models/resnet/test.py:
```diff
@@ -0,0 +1,20 @@
+seed_everything(42)
+
+
+@pytest.mark.private_access
```
```suggestion
```

mschrimpf commented on brainscore_vision/models/squeezenet/model.py:
```diff
@@ -0,0 +1,49 @@
+def get_model(net):
+    assert net in net_constructors, f"Could not find SqueezeNet network: {net}"
+
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
move to wrapper

mschrimpf commented on brainscore_vision/models/squeezenet/test.py:
```diff
@@ -0,0 +1,29 @@
+from brainscore_vision import score
+from brainscore_vision.utils import seed_everything
+
+seed_everything(42)
```
move into function / delete

mschrimpf commented on brainscore_vision/models/resnet/test.py:
```diff
@@ -0,0 +1,20 @@
+
+from brainscore_vision import score
+from brainscore_vision.utils import seed_everything
+seed_everything(42)
```
move into function/delete

mschrimpf commented on brainscore_vision/models/squeezenet/test.py:
```diff
@@ -0,0 +1,29 @@
+logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
+
+
+@pytest.mark.private_access
```
```suggestion
```

mschrimpf commented on brainscore_vision/models/vision_transformer/model.py:
```diff
@@ -0,0 +1,59 @@
+def get_model(net):
+    assert net in net_constructors, f"Could not find ViT network: {net}"
+
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
move into wrapper

mschrimpf commented on brainscore_vision/models/vision_transformer/test.py:
```diff
@@ -0,0 +1,31 @@
+from brainscore_vision.utils import seed_everything
+
+logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
+seed_everything(42)
```
move into function / delete

mschrimpf commented on brainscore_vision/models/vision_transformer/test.py:
```diff
@@ -0,0 +1,31 @@
+seed_everything(42)
+
+
+@pytest.mark.private_access
```
```suggestion
```

mschrimpf reviewed (changes requested):
High-level points:
1. Just to make sure: all models have been vetted to reproduce existing scores on the website? Assuming yes, we should include this in the tests (can just add to the parametric tests with the private scores).
2. Why do we need to seed? Everything should be deterministic
3. move `torch.device` into the wrapper

shehadak commented:
Great points, Martin.

> 1. Just to make sure: all models have been vetted to reproduce existing scores on the website? Assuming yes, we should include this in the tests (can just add to the parametric tests with the private scores).

You are correct. For models that do appear on the website, I made sure that the scores match. The most recent commits also include additional tests on private benchmarks.

> 2. Why do we need to seed? Everything should be deterministic

Seeding is necessary for models that generally use randomness during inference (e.g. untrained models with random initialization, VAEs, MC dropout), resulting in different scores even for the same benchmark run. I also think, generally, it's a good practice to seed test running for consistency. None of the models in this PR use it, so I removed seeding for this PR, but I think it still makes sense to include it in #498.

> 3. move `torch.device` into the wrapper

Great catch. Removed `torch.device` since it's handled by the wrapper anyway.

mschrimpf commented on brainscore_vision/models/crossvit_pretrained/model.py:
```diff
@@ -0,0 +1,59 @@
+def get_model():
+    # Generate Model
+    model = create_model("crossvit_18_dagger_408", pretrained=True)
+    model.eval()
```
I would also move this to the wrapper

mschrimpf commented on brainscore_vision/models/densenet/test.py:
```diff
@@ -0,0 +1,33 @@
+    # Private
+    pytest.param("densenet121", "MajajHong2015.IT-pls", approx(0.5327, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("densenet161", "MajajHong2015.IT-pls", approx(0.5445, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("densenet169", "MajajHong2015.IT-pls", approx(0.5462, abs=0.0005), marks=[pytest.mark.private_access]),
```
this is different from the website (`.548`)

mschrimpf commented on brainscore_vision/models/efficientnet/test.py:
```diff
@@ -0,0 +1,41 @@
+    pytest.param("efficientnet_b3", "MajajHong2015.IT-pls", approx(0.511, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("efficientnet_b4", "MajajHong2015.IT-pls", approx(0.514, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("efficientnet_b5", "MajajHong2015.IT-pls", approx(0.528, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("efficientnet_b6", "MajajHong2015.IT-pls", approx(0.526, abs=0.0005), marks=[pytest.mark.private_access]),
```
different from website (`.541`), is this the same model? https://www.brain-score.org/model/vision/665

mschrimpf commented on brainscore_vision/models/efficientnet/test.py:
```diff
@@ -0,0 +1,41 @@
+    pytest.param("efficientnet_b4", "MajajHong2015.IT-pls", approx(0.514, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("efficientnet_b5", "MajajHong2015.IT-pls", approx(0.528, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("efficientnet_b6", "MajajHong2015.IT-pls", approx(0.526, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("efficientnet_b7", "MajajHong2015.IT-pls", approx(0.535, abs=0.0005), marks=[pytest.mark.private_access]),
```
different from website (`.543`) https://www.brain-score.org/model/vision/863

mschrimpf commented on brainscore_vision/models/efficientnet/test.py:
```diff
@@ -0,0 +1,41 @@
+@pytest.mark.memory_intense
+@pytest.mark.parametrize("model_identifier, benchmark_identifier, expected_score", [
+    # Private
+    pytest.param("efficientnet_b0", "MajajHong2015.IT-pls", approx(0.526, abs=0.0005), marks=[pytest.mark.private_access]),
```
different from website (`.547`) https://www.brain-score.org/model/vision/662

mschrimpf commented on brainscore_vision/models/efficientnet/test.py:
```diff
@@ -0,0 +1,41 @@
+    # Private
+    pytest.param("efficientnet_b0", "MajajHong2015.IT-pls", approx(0.526, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("efficientnet_b1", "MajajHong2015.IT-pls", approx(0.538, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("efficientnet_b2", "MajajHong2015.IT-pls", approx(0.529, abs=0.0005), marks=[pytest.mark.private_access]),
```
different from website (`.547`) https://www.brain-score.org/model/vision/663

mschrimpf commented on brainscore_vision/models/efficientnet/test.py:
```diff
@@ -0,0 +1,41 @@
+    pytest.param("efficientnet_b1", "MajajHong2015.IT-pls", approx(0.538, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("efficientnet_b2", "MajajHong2015.IT-pls", approx(0.529, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("efficientnet_b3", "MajajHong2015.IT-pls", approx(0.511, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("efficientnet_b4", "MajajHong2015.IT-pls", approx(0.514, abs=0.0005), marks=[pytest.mark.private_access]),
```
different from website (`.543`) https://www.brain-score.org/model/vision/664

mschrimpf commented on brainscore_vision/models/resnet/test.py:
```diff
@@ -0,0 +1,31 @@
+@pytest.mark.memory_intense
+@pytest.mark.parametrize("model_identifier, benchmark_identifier, expected_score", [
+    # Private
+    pytest.param("resnet-18", "MajajHong2015.IT-pls", approx(0.527, abs=0.0005), marks=[pytest.mark.private_access]),
```
different from website (`.531`) https://www.brain-score.org/model/vision/620

mschrimpf commented on brainscore_vision/models/resnet/test.py:
```diff
@@ -0,0 +1,31 @@
+@pytest.mark.parametrize("model_identifier, benchmark_identifier, expected_score", [
+    # Private
+    pytest.param("resnet-18", "MajajHong2015.IT-pls", approx(0.527, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("resnet-34", "MajajHong2015.IT-pls", approx(0.516, abs=0.0005), marks=[pytest.mark.private_access]),
```
different from website (`.474`) https://www.brain-score.org/model/vision/621

mschrimpf commented on brainscore_vision/models/resnet/test.py:
```diff
@@ -0,0 +1,31 @@
+    # Private
+    pytest.param("resnet-18", "MajajHong2015.IT-pls", approx(0.527, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("resnet-34", "MajajHong2015.IT-pls", approx(0.516, abs=0.0005), marks=[pytest.mark.private_access]),
+    pytest.param("resnet-50", "MajajHong2015.IT-pls", approx(0.523, abs=0.0005), marks=[pytest.mark.private_access]),
```
different from website (`.518`) https://www.brain-score.org/model/vision/622

mschrimpf commented on brainscore_vision/models/squeezenet/test.py:
```diff
@@ -0,0 +1,29 @@
+@pytest.mark.memory_intense
+@pytest.mark.parametrize("model_identifier, benchmark_identifier, expected_score", [
+    # Private
+    pytest.param("squeezenet1_0", "MajajHong2015.IT-pls", approx(0.518, abs=0.001), marks=[pytest.mark.private_access]),
```
different from website (`.459`) https://www.brain-score.org/model/vision/618

mschrimpf commented on brainscore_vision/models/squeezenet/test.py:
```diff
@@ -0,0 +1,29 @@
+@pytest.mark.parametrize("model_identifier, benchmark_identifier, expected_score", [
+    # Private
+    pytest.param("squeezenet1_0", "MajajHong2015.IT-pls", approx(0.518, abs=0.001), marks=[pytest.mark.private_access]),
+    pytest.param("squeezenet1_1", "MajajHong2015.IT-pls", approx(0.504, abs=0.001), marks=[pytest.mark.private_access]),
```
different from website (`.457`) https://www.brain-score.org/model/vision/619

mschrimpf commented on brainscore_vision/models/vision_transformer/test.py:
```diff
@@ -0,0 +1,33 @@
+@pytest.mark.memory_intense
+@pytest.mark.parametrize("model_identifier, benchmark_identifier, expected_score", [
+    # Private
+    pytest.param("vit_b_16", "MajajHong2015.IT-pls", approx(0.438, abs=0.001), marks=[pytest.mark.private_access]),
```
different from website (`.380`) https://www.brain-score.org/model/vision/735

mschrimpf commented on brainscore_vision/models/vision_transformer/test.py:
```diff
@@ -0,0 +1,33 @@
+@pytest.mark.parametrize("model_identifier, benchmark_identifier, expected_score", [
+    # Private
+    pytest.param("vit_b_16", "MajajHong2015.IT-pls", approx(0.438, abs=0.001), marks=[pytest.mark.private_access]),
+    pytest.param("vit_b_32", "MajajHong2015.IT-pls", approx(0.482, abs=0.001), marks=[pytest.mark.private_access]),
```
different from website (`.417`) https://www.brain-score.org/model/vision/736

mschrimpf commented on brainscore_vision/models/vision_transformer/test.py:
```diff
@@ -0,0 +1,33 @@
+    # Private
+    pytest.param("vit_b_16", "MajajHong2015.IT-pls", approx(0.438, abs=0.001), marks=[pytest.mark.private_access]),
+    pytest.param("vit_b_32", "MajajHong2015.IT-pls", approx(0.482, abs=0.001), marks=[pytest.mark.private_access]),
+    pytest.param("vit_l_16", "MajajHong2015.IT-pls", approx(0.534, abs=0.001), marks=[pytest.mark.private_access]),
```
different from website (`.487`) https://www.brain-score.org/model/vision/733

mschrimpf commented on brainscore_vision/models/vision_transformer/test.py:
```diff
@@ -0,0 +1,33 @@
+    pytest.param("vit_b_16", "MajajHong2015.IT-pls", approx(0.438, abs=0.001), marks=[pytest.mark.private_access]),
+    pytest.param("vit_b_32", "MajajHong2015.IT-pls", approx(0.482, abs=0.001), marks=[pytest.mark.private_access]),
+    pytest.param("vit_l_16", "MajajHong2015.IT-pls", approx(0.534, abs=0.001), marks=[pytest.mark.private_access]),
+    pytest.param("vit_l_32", "MajajHong2015.IT-pls", approx(0.5, abs=0.001), marks=[pytest.mark.private_access]),
```
different from website (`.439`) https://www.brain-score.org/model/vision/737 (or `.454` https://www.brain-score.org/model/vision/734)

mschrimpf reviewed (changes requested):
The code all looks good (aside from a few minor suggestions from the last review), but I am concerned about the apparent mismatch in scores for some of the models. In particular, `densenet169`, all `efficientnet`s, all `resnet`s, all `squeezenet`s and all `ViT`s have substantial mismatches. What is going on there?
